aurelia/aurelia
# feat(router): add au-href

# Pull Request

## 📖 Description

This PR adds the `au-href` attribute as an alternative to `href` and `route-href`. The presence of an `au-href` attribute on an anchor will override any `href` attribute. If there's no `au-href` on the anchor, the router option `useHref` decides whether the router should use the `href` attribute instead.

```html
<a au-href="foo(bar)">foo with bar</a>
<a au-href.bind="'foo(bar)'">foo with bar</a>
<a au-href.bind="{ component: 'foo', parameters: 'bar' }">foo with bar</a>
<a au-href.bind="{ component: Foo, parameters: 'bar' }">foo with bar</a>
```
The last example currently requires that the `Foo constructor` is known within the view model. For example
```typescript
import { Foo } from './foo';

export class MyComponent {
  private readonly Foo = Foo;
}
```

### 🎫 Issues

Related to https://github.com/aurelia/aurelia/issues/307, https://github.com/aurelia/aurelia/issues/369 and https://github.com/aurelia/aurelia/issues/367. 

## 👩‍💻 Reviewer Notes

Ignore everything in `packages/router/test/e2e/**` since they are just my own test applications while working.

## 📑 Test Plan

- Add new tests for `au-href` vs `href`
- Add new tests for proper navigations with `au-href`
- Make sure existing tests pass.

## ⏭ Next Steps

- Add first version of configuration
- Review parameters in viewport instructions
- Fix `realworld`
- Fix proper sync/async for lifecycle hooks
- Add tests for the new au-nav options
- Add tests for the before navigation guard
- Finish navigation skeleton
- Finish scoped viewports review
- Add tests covering scoped viewports
- Review controller parent implementation
- Review the extension points
